flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java

```diff
@@ -996,6 +1002,88 @@ public void testCanEmitBatchOfRecords() throws Exception {
+        TaskMetricGroup taskMetricGroup =
+                UnregisteredMetricGroups.createUnregisteredTaskMetricGroup();
+
+        ResultPartitionWriter[] partitionWriters = new ResultPartitionWriter[3];
+        for (int i = 0; i < partitionWriters.length; ++i) {
+            partitionWriters[i] =
+                    new RecordOrEventCollectingResultPartitionWriter<>(
+                            new ArrayDeque<>(),
+                            new StreamElementSerializer<>(
+                                    BasicTypeInfo.INT_TYPE_INFO.createSerializer(
+                                            new ExecutionConfig())));
+            partitionWriters[i].setup();
+        }
+
+        try (StreamTaskMailboxTestHarness<Integer> testHarness =
+                new StreamTaskMailboxTestHarnessBuilder<>(
+                                OneInputStreamTask::new, BasicTypeInfo.INT_TYPE_INFO)
+                        .addInput(BasicTypeInfo.INT_TYPE_INFO)
+                        .addAdditionalOutput(partitionWriters)
+                        .setupOperatorChain(new OperatorID(), new OddEvenOperator())
+                        .chain(BasicTypeInfo.INT_TYPE_INFO.createSerializer(new ExecutionConfig()))
+                        .setOperatorFactory(SimpleOperatorFactory.of(new OddEvenOperator()))
+                        .addNonChainedOutputsCount(
+                                new OutputTag<>("odd", BasicTypeInfo.INT_TYPE_INFO), 2)
+                        .addNonChainedOutputsCount(1)
+                        .build()
+                        .chain(BasicTypeInfo.INT_TYPE_INFO.createSerializer(new ExecutionConfig()))
+                        .setOperatorFactory(SimpleOperatorFactory.of(new DuplicatingOperator()))
+                        .addNonChainedOutputsCount(1)
+                        .build()
+                        .finish()
+                        .setTaskMetricGroup(taskMetricGroup)
+                        .build()) {
+            Counter numRecordsInCounter =
+                    taskMetricGroup.getIOMetricGroup().getNumRecordsInCounter();
+            Counter numRecordsOutCounter =
+                    taskMetricGroup.getIOMetricGroup().getNumRecordsOutCounter();
+
+            final int numEvenRecords = 5;
+            final int numOddRecords = 3;
+
+            for (int x = 0; x < numEvenRecords; x++) {
+                testHarness.processElement(new StreamRecord<>(2 * x));
+            }
+
+            for (int x = 0; x < numOddRecords; x++) {
+                testHarness.processElement(new StreamRecord<>(2 * x + 1));
+            }
+            assertEquals(numOddRecords + numEvenRecords, numRecordsInCounter.getCount());
+            assertEquals(
+                    numOddRecords
+                            + (numOddRecords + numEvenRecords)
+                            + (numOddRecords + numEvenRecords) * 2,
+                    numRecordsOutCounter.getCount());
```

Comment:
I don't follow this assertion. 

Shouldn't this be:

- `(numOddRecords + numEvenRecords) * 2` from the `DuplicatingOperator`
- `numOddRecords + numEvenRecords` from the first `OddEvenOperator`
- `numOddRecords + numEvenRecords` from the second `OddEvenOperator`

So `numOddRecords + numEvenRecords` * 4 in total? 🤔  

nit: Also it would help if you could extract the components of the total expected output records count to a separate local/named variables for example:
```
            int firstOddEvenOperatorOutputs = ...;
            int secondOddEvenOperatorOutputs = ...;
            int duplicatingOperatorOutput = ...;
            assertEquals(
                    firstOddEvenOperatorOutputs + secondOddEvenOperatorOutputs + duplicatingOperatorOutput,
                    numRecordsOutCounter.getCount());
```

ditto for the multi and two input case.

Reply:
The topology graph of this test case is as follows:
![image](https://github.com/apache/flink/assets/19502505/cdae0246-da53-4860-b3c8-9f2d62dfe559)
- The first operator does not have `NonChainedOutput`.
- The second operator have two `RecordWriter With Odd Tag` as this code:`.addNonChainedOutputsCount(new OutputTag<>("odd", BasicTypeInfo.INT_TYPE_INFO), 2)`

The assertion value(`numOddRecords + (numOddRecords + numEvenRecords) + (numOddRecords + numEvenRecords) * 2`) from the following three parts:

- `numOddRecords`  from the second `OddEvenOperator`'s two `RecordWriterOutput With Odd Tag`. It is only calculated once here, which is guaranteed by the logic of `BroadcastingOutputCollector`.
- `numOddRecords + numEvenRecords` from the second `OddEvenOperator`'s `RecordWriterOutput`.
- `(numOddRecords + numEvenRecords) * 2` from the `DuplicatingOperator`

It should be noted here that:
- `OddEvenOperator` will send duplicated data to the side output and normal output. So we have the first two parts above.
```
     public void processElement(StreamRecord<Integer> element) {
            if (element.getValue() % 2 == 0) {
                output.collect(evenOutputTag, element);
            } else {
                output.collect(oddOutputTag, element);
            }
            output.collect(element);
     }
```

Reply:
> ditto for the multi and two input case.](nit: Also it would help if you could extract the components of the total expected output records count to a separate local/named variables)

This really a good suggestion.

Reply:
Thanks for the explanation. I think I understand it. One thing is not clear to me.

> The first operator does not have NonChainedOutput

So what is going on with the first operator 🤔? That first operator is trying to emit records with tags, but tags are only supported in non-chained outputs. First operator doesn't have any, and as you pointed out, `ChainedOutput` is ignoring records with tags. So the `if/else` branch in the first operator's `processElement` is effectively a dead code?

If the above is correct, is this construct officially supported behaviour of Flink? Can user create such pipelines? If not, I would either remove the first operator, or replace it with just a no-op/pass-through operator that just forwards input records.

Reply:
> If the above is correct, is this construct officially supported behaviour of Flink? Can user create such pipelines? 
Fair enough! This is indeed not a conventional pipeline, I have replaced the first operator by a `PassThroughOperator`.